### ISSUE TITLE
❌  Known Issue: Einstein for Developers does not work in Code Builder environments

### ISSUE THREAD

**Issue by AnanyaJha** [forcedotcom/Einstein-GPT-for-Developers#93]
### Summary

Code Builder users will see Einstein for Developers functionality fail when executing from a Code Builder environment. This is a regression that was introduced ~2 weeks ago due to org authentication changes.

The teams are actively investigating and we'll update this thread with an ETA on a fix asap.

### Workaround

As a workaround, please use the following steps:

- Open the integrated terminal inside Code Builder (CTRL+SHIFT)
- Replace the alias and run the following CLI command `sf org login device --set-default --alias <alias>`
- Open a new integrated terminal. Example Output:
```
=== Action Required!

Enter 05XJUQHV device code in this verification URL: https://login.salesforce.com/setup/connect
```

-  Click on the verification URL and enter the provided device code. Follow the rest of the prompts and you should be able to get started with Einstein!

⚠️ **Note:** You might need to give E4D a minute to pick up the auth. I’ve noticed if I click on something that is not E4D and then click back into E4D everything works.

⚠️ **Additional note:** The CLI will connect you to your current session, that means whichever org you are currently logged in to. If you want to connect to a different org, make sure to log out of any other orgs.

**Comment by aidanfowler** [comment]
This workaround does not work. E4D continues to show not connected to default org after completing the authentication via command line.

**Comment by randi274** [comment]
Hi @aidanfowler are you seeing any additional errors in the Developer Tools, or under Output > Einstein for Developers? Here's a screen capture of a flow that worked for me to log in using the device login. I also been able to work with scratch orgs if I set up a Dev Hub and then create a scratch org from my dev hub.

https://github.com/user-attachments/assets/8ce8219a-fc73-492a-a0da-b4fbdccabda7

**Comment by aidanfowler** [comment]
After repeating your steps, there is an unknown error popup in the bottom right corner. Dev assistant shows "The Salesforce org you're connected to can't access Einstein for Developers. Connect to a different org and try again. Log an issue if the problem persists.". 

Uploading Screen Recording 2024-09-09 at 11.49.33 AM.mov…

**Comment by randi274** [comment]
Thank you for adding the video as well! Is it possible for you to run `>SFDX: Log Out from all Authorized Orgs` first? I realize now I had done that first, restarted Code Builder, and then started with my auth flow.

**Comment by aidanfowler** [comment]
@randi274 that worked, thanks so much!

**Comment by thedanfernandez** [comment]
Hey,

@randi274, I tried to reset my environment, and logged out of all orgs but I'm getting a Type error, below: 

1. List all orgs returns no orgs found

```bash
$ sf org list --all
 ›   Warning: @salesforce/cli update available from 2.57.7 to 2.58.7.
```

**Result**
```
No Orgs found.
```

2. Login to an org (not a scratch org)
```bash
$ sf org login device --set-default --alias myorg
 ›   Warning: @salesforce/cli update available from 2.57.7 to 2.58.7.
=== Action Required!

Enter XXXXXX device code in this verification URL: https://login.salesforce.com/setup/connect
```

**Result**
```
    TypeError: Cannot read properties of undefined (reading 'includes')
```

**Comment by goravseth** [comment]
I have tried the workarounds above multiple times, including logout from all orgs, but get the same 'oops looks like you ran into an issue with running E4D' error message with a link to this page.

If I refresh the browser window, E4D appears to load for a sec, and i can see the input prompt, and then it quickly gets replaced with "The Salesforce org you're connected to can't access Einstein for Developers."...  i've tried prod and a partial-copy sandbox.

any ETA on a fix?  or other suggestions to try?

**Comment by goravseth** [comment]
Tried again and was finally successful.  This time I connected to prod first.  

However, once I switched my default org to my partial copy sandbox, I got the error "The Salesforce org you're connected to can't access Einstein for Developers. Connect to a different org and try again"

Using a different sandbox is working.  Pehrpas e4d doesnt work with partial copy sandboxes, or perhaps the sandbox needs to be refreshed.  Calling it good for now.

**Comment by dstewart-fresenius** [comment]
I am continually running into the error, lower left, indicating 'No default org set.' after logging out of all orgs, restarting Code Builder and running the 'sf org login device --set-default --alias <default>' command.   Question: Should the default be my production org, where Einstein is enabled, since that's where login.salesforce.com/.. takes me?

I am seeing the 'Login successful for <username>' in Terminal, and the Einstein: Dev Assistant looks hopeful for a moment, then it refreshes back to 'Welcome!..' and 'Connect to a Salesforce Org' and the little brown error in the lower right reappears w' 'No Default Org set.'

**Comment by gbockus-sf** [comment]
This issue has been resolved and you no longer have to do the work around to get A4D access in Codebuilder.

**Comment by gbockus-sf** [comment]
If you get the situation where you have auth'd but A4D isn't showing up as enabled: 
1. Be aware of odd known issue where if you use the same alias on multiple authentication attempts in the same session A4D will not be able to detect the login that succeeded last #85 
2. If not the same alias issue please open a new issue as this is a new condition and we'll need to dig in to figure out what is causing the issue. 

Thanks!

**Comment by Vikastnsi** [comment]
@gbockus-sf @randi274 
I have tried the workaround  but the work around is not working for me

Workaround:
1) Open the 'Code Builder' from production and launch it.
2) Open the integrated terminal inside Code Builder (CTRL+SHIFT)
3) Replace the alias and run the following CLI command sf org login device --set-default --alias <alias>
4) Getting waring message: Warning: @salesforce/cli update available from 2.28.6 to 2.71.6.
5) Open a new integrated terminal. Example Output:
=== Action Required!

Enter XXXXXXX device code in this verification URL: https://login.salesforce.com/setup/connect

6) Opened the above link and entered the device code  and submit it.
7) 'Agent-force for the developer' still not logged in Org
